lib/netext/dialer.go

```diff
@@ -149,6 +141,66 @@ func (d *Dialer) GetTrail(
 	}
 }
 
+func (d *Dialer) dialAddr(addr string) (string, error) {
+	remote, err := d.findRemote(addr)
+	if err != nil {
+		return "", err
+	}
+
+	for _, ipnet := range d.Blacklist {
+		if (*net.IPNet)(ipnet).Contains(remote.IP) {
+			return "", BlackListedIPError{ip: remote.IP, net: ipnet}
+		}
+	}
+
+	return remote.String(), nil
+}
+
+func (d *Dialer) findRemote(addr string) (*lib.HostAddress, error) {
+	host, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	remote, err := d.preloadRemote(addr, host, port)
+	if err != nil || remote != nil {
+		return remote, err
+	}
+
+	ip := net.ParseIP(host)
+
+	if ip == nil {
+		ip, err = d.Resolver.FetchOne(host)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if ip == nil {
+		return nil, errors.Errorf("lookup %s: no such host", host)
+	}
+
+	return lib.NewHostAddress(ip, port)
+}
+
+func (d *Dialer) preloadRemote(addr, host, port string) (*lib.HostAddress, error) {
+	// lookup for full address defined in Hosts option before trying to resolve DNS.
+	remote := d.Hosts[addr]
+	if remote == nil {
+		// lookup for host defined in Hosts option before trying to resolve DNS.
+		remote = d.Hosts[host]
+	}
+
+	if remote != nil && remote.Port == 0 && port != "" {
```

Comment:
won't it be easier to  be just 

```
if remote == nil {
return nil,nil
}
```

Also I am not really certain the name is accurate ... maybe `checkInHosts`? and maybe it should return a bool which is `true` if the check was possitive

Reply:
Also looking at this should this whole function be 
```
if remote, ok :=d.Hosts[addr]; ok{
  return remote, nil
}
if remote, ok := d.Hosts[host]; ok {
// here is the part which adds the port , which also could possibly be a integer here ? so you don't need to have a change for an error ?
}
return nil,nil
```
Because currently even if the whole `addr` matches you will make a new HostAddress ?